File: app/src/main/java/hibernate/v2/testyourandroid/ui/fragment/HardwareNFCFragment.java
package hibernate.v2.testyourandroid.ui.fragment;

import android.app.PendingIntent;
import android.content.Intent;
import android.content.IntentFilter;
import android.nfc.NfcAdapter;
import android.nfc.tech.NfcF;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;

import butterknife.ButterKnife;
import hibernate.v2.testyourandroid.C;
import hibernate.v2.testyourandroid.R;

/**
 * Created by himphen on 21/5/16.
 */
public class HardwareNFCFragment extends BaseFragment {

	private NfcAdapter mAdapter;
	private PendingIntent mPendingIntent;
	private IntentFilter[] mIntentFilters;
	private String[][] mNFCTechLists;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setHasOptionsMenu(true);
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		// Inflate the layout for this fragment
		View rootView = inflater.inflate(R.layout.fragment_hardware_nfc, container, false);
		ButterKnife.bind(this, rootView);
		return rootView;
	}

	@Override
	public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		init();
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_settings:
				C.startSettingsActivity(mContext, Settings.ACTION_NFC_SETTINGS);
				break;
		}
		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onPause() {
		super.onPause();
		if (mAdapter != null)
			mAdapter.disableForegroundDispatch(mContext);
	}

	@Override
	public void onResume() {
		super.onResume();
		if (mAdapter != null)
			mAdapter.enableForegroundDispatch(mContext, mPendingIntent,
					mIntentFilters, mNFCTechLists);
	}

	private void init() {

		mAdapter = NfcAdapter.getDefaultAdapter(mContext);
		if (mAdapter != null) {
			if (!mAdapter.isEnabled()) {
				openNFCDialog();
			}
		} else {
			C.errorNoFeatureDialog(mContext);
		}

		// create an intent with tag data and deliver to this activity
		mPendingIntent = PendingIntent.getActivity(mContext, 0, new Intent(mContext,
				getClass()).addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), 0);

		// set an intent filter for all MIME data
		IntentFilter ndefIntent = new IntentFilter(
				NfcAdapter.ACTION_NDEF_DISCOVERED);
		try {
			ndefIntent.addDataType("*/*");
			mIntentFilters = new IntentFilter[]{ndefIntent};
		} catch (Exception e) {
			Log.e("TagDispatch", e.toString());
		}

		mNFCTechLists = new String[][]{new String[]{NfcF.class.getName()}};
	}

	public void openNFCDialog() {
		MaterialDialog.Builder dialog = new MaterialDialog.Builder(mContext)
				.title(R.string.ui_caution)
				.content(R.string.nfc_enable_message)
				.positiveText(R.string.nfc_enable_posbtn)
				.onPositive(new MaterialDialog.SingleButtonCallback() {
					@Override
					public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
						C.startSettingsActivity(mContext, Settings.ACTION_NFC_SETTINGS);
					}
				})
				.negativeText(R.string.ui_cancel);
		dialog.show();
	}
}

File: app/src/main/java/hibernate/v2/testyourandroid/ui/fragment/HardwareNFCFragment.kt
package hibernate.v2.testyourandroid.ui.fragment

import android.app.PendingIntent
import android.content.Intent
import android.content.IntentFilter
import android.nfc.NfcAdapter
import android.nfc.tech.NfcF
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.afollestad.materialdialogs.MaterialDialog
import hibernate.v2.testyourandroid.R
import hibernate.v2.testyourandroid.helper.UtilHelper.errorNoFeatureDialog
import hibernate.v2.testyourandroid.helper.UtilHelper.startSettingsActivity

/**
 * Created by himphen on 21/5/16.
 */
class HardwareNFCFragment : BaseFragment() {
    private var mAdapter: NfcAdapter? = null
    private var mPendingIntent: PendingIntent? = null
    private lateinit var mIntentFilters: Array<IntentFilter>
    private lateinit var mNFCTechLists: Array<Array<String>>
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_hardware_nfc, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> startSettingsActivity(context, Settings.ACTION_NFC_SETTINGS)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()
        if (mAdapter != null) mAdapter!!.disableForegroundDispatch(activity)
    }

    override fun onResume() {
        super.onResume()
        if (mAdapter != null) mAdapter!!.enableForegroundDispatch(activity, mPendingIntent,
                mIntentFilters, mNFCTechLists)
    }

    private fun init() {
        mAdapter = NfcAdapter.getDefaultAdapter(context)
        if (mAdapter != null) {
            if (!mAdapter!!.isEnabled) {
                openFunctionDialog()
            }
        } else {
            errorNoFeatureDialog(context)
        }
        // create an intent with tag data and deliver to this activity
        mPendingIntent = PendingIntent.getActivity(context, 0, Intent(context,
                javaClass).addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), 0)
        // set an intent filter for all MIME data
        val ndefIntent = IntentFilter(
                NfcAdapter.ACTION_NDEF_DISCOVERED)
        try {
            ndefIntent.addDataType("*/*")
            mIntentFilters = arrayOf(ndefIntent)
        } catch (e: Exception) {
            Log.e("TagDispatch", e.toString())
        }
        mNFCTechLists = arrayOf(arrayOf(NfcF::class.java.name))
    }

    private fun openFunctionDialog() {
        context?.let {
            MaterialDialog(it)
                    .title(R.string.ui_caution)
                    .message(R.string.nfc_enable_message)
                    .cancelable(false)
                    .positiveButton(R.string.nfc_enable_posbtn) { dialog ->
                        startSettingsActivity(context, Settings.ACTION_NFC_SETTINGS)
                    }
                    .negativeButton(R.string.ui_cancel)
                    .show()
        }
    }
}